### ISSUE TITLE
Day 3 review: Releasing Code

### ISSUE THREAD

**Issue by mwhosek** [semaphoreP/codeastro#105]
I think the slides do a great job of going through how to release code! I also think the material on how to organize your repository (with `__init__.py`, sub-packages, etc) is super useful, and could be a topic in its own right. I could see this lesson organized into 2 parts: (1) how to organize your code repository, and then (2) how to release code.

I had a few small comments on the slides:

1. Slide 6: it might be helpful to add a bit more info regarding the dotted naming scheme, and how one can use it to import modules. For example, in the example in this slide, utils is a sub-package of orbitize, right? Perhaps you could include an example of how the dotted naming scheme would work on the slide 7 (the `sound/` package)...for example, I think you'd import `echo` as 

`from sound.effects import echo`

To me, the dotted naming scheme makes more sense when I see how the files in the repository are organized.

2. Slide 12 (`__init__.py`): It might be useful to include a minimal example of an `__init__.py` file, so folks get an idea of what it looks like. These can be blank in principle, right? I personally don't know too much about them.

3. Slide 21 (zenodo): I recently learned that you can link your zenodo account to your GitHub repository such that zenodo will automatically create a release (with a unique DOI) every time you create a GitHub release. The procedure for doing this is fairly straightforward (instructions [here](https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content)). Is this worth talking about (maybe as extra material)? It is convenient to have zenodo just do its thing automatically whenever you do a GitHub release. 

Finally, I had a question myself regarding the material:
When you install something locally via pip install -e, where does it go? For example, normally I need to have my `PYTHONPATH` variable pointing to a code repository in order to run the package. If I instead `pip install` the package, then is the `PYTHONPATH` no longer needed?

Thanks!

**Comment by mwhosek** [comment]
And a follow-up to my question: if I `pip install` a package, and then make changes to that package in the repository (without redoing the `pip install`), which version of the code gets used when I try to use it? The pip install version, or the repository version (assuming the repository is in my PYTHONPATH).

Anyway, this probably isn't important for the workshop, but I am curious :).

**Comment by sblunt** [comment]
Hey Matt, thanks so much for the excellent suggestions! I've implemented the changes you suggested in the google slides for this lesson.

Re: your question about pip-- check out the `pip show` command. If I run `pip show orbitize` in an environment where I've got it installed, this is the output:

```
Name: orbitize
Version: 2.1.1
Summary: orbitize! Turns imaging data into orbits
Home-page: https://github.com/sblunt/orbitize
Author: 
Author-email: 
License: BSD
Location: /Users/bluez3303/miniconda3/envs/python3.8/lib/python3.8/site-packages
Requires: astropy, astroquery, corner, cython, deprecation, docutils, emcee, h5py, matplotlib, numpy, pandas, ptemcee, pyerfa, pytest, radvel, rebound, scipy, sphinx
Required-by: 
```

Here's my understanding/explanation of what's going on:

When you activate a conda environnment, the $PATH is modified so that the first place python looks for packages is within that conda environment's "storage" location. In my current conda environment (called python3.8), you can see that executables for all python packages get stored in ` /Users/bluez3303/miniconda3/envs/python3.8/lib/python3.8/site-packages`. `pip install -e` creates an executable based on the version of the source code you have on your machine, and tells pip that whenever you save the code, it should update that executable (thus propagating your changes). 

Note that conda operates on the $PATH environment variable, not the $PYTHONPATH environment variable, so it can get dicey trying to use both of them at the same time. Personally, I don't use PYTHONPATH, and use a combination of conda environments and explicit $PATH additions to manage the imports I need. I'm not sure what would happen in the particular situation you described, but here's a few examples of what would happen in various situations (trying to answer your question):

1. I have git cloned `orbitize` to my desktop. Within the orbitize directory, I run:
`pip install orbitize`
This tells pip to query pypi for packages named orbitize, so it will totally ignore the fact that I have a local copy of a package named orbitize, and will create an executable using the version on pypi. 

2. I have git cloned `orbitize` to my desktop, and run `pip install orbitize`, as above. Within the orbitize directory, I now run:
`pip install .`
I wasn't sure what would happen here, but it turned out that pip overwrote my previous installation based on the pypi version of orbitize, and created an executable based on my local copy of the code.  However, since I didn't include the `-e` option, which is what tells pip to update the executable every time I save the code, any new changes I make won't be reflected in the executable.
